Add support for Velocity IP forwarding

astei opened the pull request:
While Velocity already supports BungeeCord-style IP forwarding (which almost everyone supports now), it is not secure. Users have a lot of problems setting up firewalls or setting up plugins like iPWhitelist. Further, the BungeeCord IP forwarding protocol still retains essentially its original form, when there is brand new support for custom login plugin messages in 1.13.

Velocity's modern IP forwarding uses an HMAC-SHA256 code to ensure authenticity of messages (users have to set up an unique secret), is packed into a binary format that is smaller than BungeeCord's forwarding, and is integrated into the Minecraft login process by using the 1.13 login plugin message packet.

I have verified this with Velocity build 303 in online-mode.

aikar commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
+             }
+         }
+     }
++    public String readUtf(int maxLength) { return this.e(maxLength); } // Paper - OBFHELPER
```
move down a line, change to UTF

aikar commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
+     public UUID i() {
+         return new UUID(this.readLong(), this.readLong());
+     }
++    public UUID readUuid() { return this.i(); } // Paper - OBFHELPER
```
move down a line, change to UUID

aikar commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
+ 
+         return i;
+     }
++    public int readVarInt() { return this.g(); } // Paper - OBFHELPER
```
move down a line

aikar commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
+         });
+ 
+         if (minecraftserver.getOnlineMode()
++                || (com.destroystokyo.paper.PaperConfig.velocitySupport && com.destroystokyo.paper.PaperConfig.velocityOnlineMode) // Paper - Velocity support
```
We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()

astei replied:
I was considering that, but wondered if we wanted a separate patch for that.

aikar replied:
We can put in the stub method in the bungee patch, then velocity amends that.

astei replied:
Will do this tonight.

aikar commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
+ 
+         version = getInt("config-version", 14);
+-        set("config-version", 15);
++        set("config-version", 16);
```
Not needed. Just add getXX calls in the method and itll add the defaults for you.

aikar commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
++    public static boolean velocityOnlineMode;
++    public static byte[] velocitySecretKey;
++    private static void velocitySupport() {
++        if (version < 16) {
```
as said above, all the below getXX methods will set these defaults for you. versions only needed for migration paths, not new keys.

aikar commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
+     public boolean complete(boolean textures) {
+         MinecraftServer server = MinecraftServer.getServer();
+ 
+-        boolean isOnlineMode = server.getOnlineMode() || (SpigotConfig.bungee && PaperConfig.bungeeOnlineMode);
```
We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()

aikar commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
+         UserCache userCache = server.getUserCache();
+         if (profile.getId() == null) {
+             final GameProfile profile;
+-            boolean isOnlineMode = server.getOnlineMode() || (SpigotConfig.bungee && PaperConfig.bungeeOnlineMode);
```
We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()

aikar commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
+         GameProfile profile;
+         // Only fetch an online UUID in online mode
+         if (MinecraftServer.getServer().getOnlineMode()
+-                || (org.spigotmc.SpigotConfig.bungee && com.destroystokyo.paper.PaperConfig.bungeeOnlineMode)) {
```
We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()

aikar commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
+             GameProfile profile = null;
+             // Only fetch an online UUID in online mode
+             if ( MinecraftServer.getServer().getOnlineMode()
++                    || (com.destroystokyo.paper.PaperConfig.velocitySupport && com.destroystokyo.paper.PaperConfig.velocityOnlineMode) // Paper - Velocity support
```
We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()

kashike commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
+     private EntityPlayer l;
+     public String hostname = ""; // CraftBukkit - add field
++    // Paper start - Velocity support
++    private static final MinecraftKey VELOCITY_CHANNEL = new MinecraftKey("velocity", "player_info");
```
`VELOCITY_PLAYER_INFO_CHANNEL`

electronicboy commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
++        velocityOnlineMode = getBoolean("settings.velocity-support.online-mode", false);
++        String secret = getString("settings.velocity-support.secret", "");
++        if (velocitySupport && secret.isEmpty()) {
++            logError("Velocity support is enabled, but no secret key was specified. Velocity support will not be enabled.");
```
Some part of me feels that if we're going to support this, this should be more critical

astei replied:
What would be the best way to make this fatal? Perhaps `System.exit(1)`?

kashike commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,294 @@
+     // Spigot end
+ 
+     public void a(PacketLoginInCustomPayload packetloginincustompayload) {
++        // Paper start - Velocity support
```
I think all the logic that can be extracted should be extracted to its own class: `io.papermc.paper.proxy.VelocityProxy` or so.

astei replied:
The only code I think we can extract is the logic that begins the login process in the thread pool executor, which is repeated three times in this file in this PR (and two before).

kashike commented on Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,299 @@
+     private EntityPlayer l;
+     public String hostname = ""; // CraftBukkit - add field
++    // Paper start - Velocity support
++    private static final MinecraftKey VELOCITY_PLAYER_INFO_CHANNEL = new MinecraftKey("velocity", "player_info");
```
Can be moved to https://github.com/PaperMC/Paper/pull/1557/files#r223475928

kashike commented on Spigot-Server-Patches/0390-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,308 @@
+         Bukkit.getLogger().severe(s);
+     }
+ 
++    protected static void fatal(String s) {
```
This fatal stuff should be moved to the root config patch.

zachbr commented on Spigot-Server-Patches/0390-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,308 @@
++                return;
++            }
++
++            this.networkManager.l = new java.net.InetSocketAddress(com.destroystokyo.paper.proxy.VelocityProxy.readAddress(buf), ((java.net.InetSocketAddress) this.networkManager.getSocketAddress()).getPort());
```
Add an obfuscation helper as a setter right next to the field declaration please

kashike commented on Spigot-Server-Patches/0390-Add-Velocity-IP-Forwarding-Support.patch:
```diff
@@ -0,0 +1,300 @@
+ import com.destroystokyo.paper.profile.PlayerProfile;
+ import com.mojang.authlib.GameProfile;
+ import com.mojang.authlib.exceptions.AuthenticationUnavailableException;
+-import io.netty.channel.ChannelFuture;
```
Don't touch imports.

astei committed:
Add Velocity IP forwarding support

While Velocity supports BungeeCord-style IP forwarding, it is not secure. Users
have a lot of problems setting up firewalls or setting up plugins like IPWhitelist.
Further, the BungeeCord IP forwarding protocol still retains essentially its original
form, when there is brand new support for custom login plugin messages in 1.13.

Velocity's modern IP forwarding uses an HMAC-SHA256 code to ensure authenticity
of messages, is packed into a binary format that is smaller than BungeeCord's
forwarding, and is integrated into the Minecraft login process by using the 1.13
login plugin message packet.
